File: zmy/golang/2.add-two-numbers.go
/*
 * @lc app=leetcode id=2 lang=golang
 *
 * [2] Add Two Numbers
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	head := &ListNode{}
	p := head
	acc := 0
	for l1 != nil || l2 != nil || acc !=0 {
		if l1 != nil {
			acc += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			acc += l2.Val
			l2 = l2.Next
		}
		p.Next = &ListNode{Val: acc % 10, Next: nil}
		acc /= 10
		p = p.Next
	}
	return head.Next
}

File: zmy/cpp/2.add-two-numbers.cpp
/*
 * @lc app=leetcode id=2 lang=cpp
 *
 * [2] Add Two Numbers
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *head = new ListNode(0);
        ListNode *p = head;
        int acc = 0;
        while (l1 != nullptr || l2 != nullptr || acc != 0) {
            p->next = new ListNode(0);
            p = p->next;
            if (l1 != nullptr) {
                acc += l1->val;
                l1 = l1->next;
            }
            if (l2 != nullptr) {
                acc += l2->val;
                l2 = l2->next;
            }
            p->val = acc % 10;
            acc /= 10;
        }
        return head -> next;

    }
};

